README.rst
**********************
NEMO Command Processor
**********************

The NEMO command processor, ``nemo``, is a command line tool for doing various operations associated with running the `NEMO`_ ocean model.

.. _NEMO: http://www.nemo-ocean.eu/

Use ``nemo --help`` to get a list of the sub-commands available.
Use ``nemo help <sub-command>`` to get a synopsis of what a sub-command does,
what its required arguments are,
and what options are available to control it.

Documentation for the command processor is in the ``docs/`` directory and is rendered at https://nemo-cmd.readthedocs.io/en/latest/.

.. image:: https://readthedocs.org/projects/nemo-cmd/badge/?version=latest
    :target: https://nemo-cmd.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

This an extensible tool built on the OpenStack ``cliff``
(`Command Line Interface Formulation Framework`_)
package.
As such,
it can be used as the basis for a NEMO domain-specific command processor tool.

.. _Command Line Interface Formulation Framework: http://docs.openstack.org/developer/cliff/

The ``NEMO-Cmd`` is based on v2.2 of the Salish Sea MEOPAR NEMO model project's ``tools/SalishSeaCmd`` package.


License
=======

The NEMO command processor and documentation are copyright 2013-2017 by the Salish Sea MEOPAR Project Contributors and The University of British Columbia.

They are licensed under the Apache License, Version 2.0.
https://www.apache.org/licenses/LICENSE-2.0
Please see the LICENSE file for details of the license.


docs/CHANGES.rst
**********
Change Log
**********

Next Release
============

* Expand shell and user variables in namelist file paths.

* Use resolved repo path in VCS revisions recording message about uncommitted
  changes.

* Change to copy ref namelists to temporary run dir instead of symlinking them;
  facilitates easier run result archeology and reproducibility.

* Fix bug in atmospheric forcing file links checking function call.


1.0
===

* Enable ``namelist.namelist2dict()`` to handle Fortran boolean values ``true``
  and ``false`` (no leading/trailing dots).

* Confirm that the ``rebuild_nemo.exe`` executable in the ``prepare`` plug-in
  so that a run is not executed without it only to fail when the ``combine``
  plug-in is run.
  See https://bitbucket.org/salishsea/nemo-cmd/issues/19.

* Add find_rebuild_nemo_script() to the API.
  See https://bitbucket.org/salishsea/nemo-cmd/issues/20.

* For NEMO-3.6 only,
  restart file paths/filenames are now specified in a new ``restart`` section
  instead of in the ``forcing`` section.
  See https://nemo-cmd.readthedocs.io/en/latest/run_description_file/3.6_yaml_file.html#restart-section.

* The existence of all paths/files given in the run description YAML file
  is confirmed.
  An informative error message is emitted for paths/files that don't exist.

* Add optional ``filedefs`` item to output section of run description YAML
  file to facilitate the use of a ``file_Def.xml`` file with XIOS-2.

* Change spelling of keys in output section of run description YAML file:

  * ``files`` becomes ``iodefs``
  *  ``domain`` becomes ``domaindefs``
  *  ``fields`` becomes ``fielddefs``

  Old spellings are retained as fall-backs for backward compatibility.

* Fix Python 2.7 Unicode/str issue in Mercurial version control revision
  and status recording.
  See https://bitbucket.org/salishsea/nemo-cmd/issues/16.

* Add option to provide in the run description YAML file a list of
  PBS resource key-value pairs to produce ``#PBS -l`` directives for in the
  run shell script.
  See https://nemo-cmd.readthedocs.io/en/latest/run_description_file/3.6_yaml_file.html#pbs-resources-section,
  and https://bitbucket.org/salishsea/nemo-cmd/issues/10.

* Add option to provide in the run description YAML file a list of
  HPC environment modules to include ``module load`` commands for in the
  run shell script.
  See https://nemo-cmd.readthedocs.io/en/latest/run_description_file/3.6_yaml_file.html#modules-to-load-section,
  and https://bitbucket.org/salishsea/nemo-cmd/issues/11.

* Add the option to use absolute paths for coordinates and bathymetry files
  in the run description YAML file.
  See https://nemo-cmd.readthedocs.io/en/latest/run_description_file/3.6_yaml_file.html#grid-section,
  and https://bitbucket.org/salishsea/nemo-cmd/issues/5.

* Add ``nemo_cmd.fspath()``,
  ``nemo_cmd.expanded_path()``,
  and ``nemo_cmd.resolved_path()`` functions for
  working with file system paths.
  See https://nemo-cmd.readthedocs.io/en/latest/api.html#functions-for-working-with-file-system-paths.

* Port in the SalishSeaCmd ``run`` plug-in in a minimal form sufficient for
  use on TORQUE/PBS systems that don't require special PBS feature (-l)
  directives,
  or loading of environment modules.

* Add optional recording of revision and status of Mercurial version control
  repositories via a new ``vcs revisions`` section in the run description YAML
  file.

* For NEMO-3.6 only,
  enable the use of ref namelists from directories other than from
  ``CONFIG/SHARED/``.
  The default is to symlink to ``CONFIG/SHARED/namelist*_ref`` when there are no
  ``namelist*_ref`` keys in the ``namelists`` section of the run description
  YAML file.

* Change from using pathlib to pathlib2 package for Python 2.7 because the
  latter is the backport from the Python 3 stdlib that is being kept up to date.

* Refactor the ``combine`` plug-in to only run ``rebuild_nemo`` to combine
  per-processor results and/or restart files.

* Add ``deflate`` plug-in to deflate variables in netCDF files using Lempel-Ziv
  compression.

* Fix a bug whereby results directories were gathered with a redundant directory
  layer;
  e.g. the files in ``runs/9e5958d4-cb95-11e6-a99b-00259059edac/restart/``
  were gathered to ``results/25dec16/restart/restart/`` instead of
  ``results/25dec16/restart/``.


0.9
===

* Use `tox`_ for unified Python 2.7 and 3.5 testing.

  .. _tox: https://tox.readthedocs.io/en/latest/

* Refactor the ``gather`` plug-in in a minimal form sufficient for use by the
  ``GoMSS_Nowcast`` package.

* Refactor the ``prepare`` plug-in as the first ``nemo`` subcommand.

* Add token-based Fortran namelist parser from gist.github.com/krischer/4943658.
  That module also exists in the ``tools/SalishSeaTools`` package.
  It was brought into this package to avoid making this package depend on
  ``SalishSeaTools``.

* Adopt yapf for code style management.
  Project-specific style rules are set in ``.style.yapf``.

* Initialize project from the SalishSeaCmd/ directory of the tools repo with::

    hg convert --filemap tools/NEMO-Cmd_filemap.txt tools NEMO-Cmd

  A copy of ``NEMO-Cmd_filemap.txt`` is included in this repo.


nemo_cmd/api.py
# Copyright 2013-2017 The Salish Sea MEOPAR Contributors
# and The University of British Columbia

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#    http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""NEMO command processor API

Application programming interface for the NEMO command processor.
Provides Python function interfaces to command processor sub-commands
for use in other sub-command processor modules,
and by other software.
"""
import datetime
import logging
import os
try:
    from pathlib import Path
except ImportError:
    # Python 2.7
    from pathlib2 import Path
import subprocess

import cliff.commandmanager
import yaml

from nemo_cmd import combine as combine_plugin
from nemo_cmd import deflate as deflate_plugin
from nemo_cmd import gather as gather_plugin
from nemo_cmd import prepare as prepare_plugin

log = logging.getLogger(__name__)
log.setLevel(logging.DEBUG)
handler = logging.StreamHandler()
handler.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(name)s %(levelname)s: %(message)s')
handler.setFormatter(formatter)
log.addHandler(handler)


def combine(run_desc_file):
    """Run the NEMO :program:`rebuild_nemo` tool for each set of
    per-processor results files.

    The output of :program:`rebuild_nemo` for each file set is logged
    at the INFO level.

    :param run_desc_file: File path/name of the run description YAML file.
    :type results_dir: :py:class:`pathlib.Path`
    """
    return combine_plugin.combine(run_desc_file)


def deflate(filepaths, max_concurrent_jobs):
    """Deflate variables in each of the netCDF files in filepaths using
    Lempel-Ziv compression.

    Converts files to netCDF-4 format.
    The deflated file replaces the original file.

    :param sequence filepaths: Paths/names of files to be deflated.

    :param int max_concurrent_jobs: Maximum number of concurrent deflation
                                    processes allowed.
    """
    try:
        return deflate_plugin.deflate(filepaths, max_concurrent_jobs)
    except AttributeError:
        # filepaths is sequence of path strings not Path objects
        return deflate_plugin.deflate(
            map(Path, filepaths), max_concurrent_jobs
        )


def find_rebuild_nemo_script(run_desc):
    """Calculate absolute path of the rebuild_nemo script.

    Confirm that the rebuild_nemo executable exists, raising a SystemExit
    exception if it does not.

    :param dict run_desc: Run description dictionary.

    :return: Resolved path of :file:`rebuild_nemo` script.
    :rtype: :py:class:`pathlib.Path`

    :raises: :py:exc:`SystemExit` if the :file:`rebuild_nemo` script does not
             exist.
    """
    return combine_plugin.find_rebuild_nemo_script(run_desc)


def gather(results_dir):
    """Move all of the files and directories from the present working directory
    into results_dir.

    If results_dir doesn't exist, create it.

    Delete any symbolic links so that the present working directory is empty.

    :param results_dir: Path of the directory into which to store the run
                        results.
    :type results_dir: :py:class:`pathlib.Path`
    """
    return gather_plugin.gather(results_dir)


def prepare(run_desc_file, nemo34=False, nocheck_init=False):
    """Prepare a NEMO run.

    A UUID named temporary run directory is created and symbolic links
    are created in the directory to the files and directories specifed
    to run NEMO.
    The output of :command:`hg parents` is recorded in the directory
    for the NEMO-code and NEMO-forcing repos that the symlinks point to.
    The path to the run directory is returned.

    :param run_desc_file: File path/name of the YAML run description file.
    :type run_desc_file: :py:class:`pathlib.Path`

    :arg boolean nemo34: Prepare a NEMO-3.4 run;
                         the default is to prepare a NEMO-3.6 run

    :arg nocheck_init: Suppress initial condition link check the
                       default is to check
    :type nocheck_init: boolean

    :returns: Path of the temporary run directory
    :rtype: :py:class:`pathlib.Path`
    """
    return prepare_plugin.prepare(run_desc_file, nemo34, nocheck_init)


def run_description(
    config_name='SalishSea',
    run_id=None,
    walltime=None,
    mpi_decomposition='8x18',
    NEMO_code=None,
    XIOS_code=None,
    forcing_path=None,
    runs_dir=None,
    forcing=None,
    init_conditions=None,
    namelists=None,
    nemo34=False
):
    """Return a NEMO run description dict template.

    Value may be passed for the keyword arguments to set the value of the
    corresponding items.
    Otherwise,
    the returned run description dict
    that must be updated by assignment statements to provide those values.

    .. note::

        The value of the :kbd:`['forcing']['atmospheric']` item is set to
        :file:`/results/forcing/atmospheric/GEM2.5/operational/` which is
        appropriate for runs on :kbd:`salish`, but needs to be changed for runs
        on WestGrid.

    :arg str config_name: NEMO configuration name to use for the run.

    :arg str run_id: Job identifier that appears in the :command:`qstat`
                     listing.

    :arg str walltime: Wall-clock time requested for the run.

    :arg str mpi_decomposition: MPI decomposition to use for the run.

    :arg str NEMO_code: Path to the :file:`NEMO-code/` directory where the
                        NEMO executable, etc. for the run are to be found.
                        If a relative path is used it will start from the
                        current directory.

    :arg str XIOS_code: Path to the :file:`XIOS/` directory where the
                        XIOS executable for the run are to be found.
                        If a relative path is used it will start from the
                        current directory.

    :arg str forcing_path: Path to the :file:`NEMO-forcing/` directory
                           where the netCDF files for the grid coordinates,
                           bathymetry, initial conditions, open boundary
                           conditions, etc. are found.
                           If a relative path is used it will start from
                           the current directory.

    :arg str runs_dir: Path to the directory where run directories will be
                       created.
                       If a relative path is used it will start from the
                       current directory.

    :arg dict forcing: Forcing link data structure.
                       The default of :py:obj:`None` produces "sensible
                       defaults" for NEMO-3.4,
                       but :py:obj:`None` for NEMO-3.6.
                       See the :ref:`RunDescriptionFileStructure` docs
                       for the version of NEMO that you are using for
                       details of the data structure.

    :arg str init_conditions: Name of sub-directory in :file:`NEMO-forcing/`
                              where initial conditions files are to be found,
                              or the path to and name of a restart file.
                              If a relative path is used for a restart file
                              it will start from the  current directory.

    :arg dict namelists: Namelists data structure.
                         The default of :py:obj:`None` produces "sensible
                         defaults" for a physics-only run for both NEMO-3.4,
                         annd NEMO-3.6.
                         See the :ref:`RunDescriptionFileStructure` docs
                         for the version of NEMO that you are using for
                         details of the data structure.

    :arg boolean nemo34: Return run description dict template a NEMO-3.4
                         run;
                         the default is to return the dict for a
                         NEMO-3.6 run.
    """
    run_description = {
        'config_name': config_name,
        'MPI decomposition': mpi_decomposition,
        'run_id': run_id,
        'walltime': walltime,
        'paths': {
            'NEMO-code': NEMO_code,
            'forcing': forcing_path,
            'runs directory': runs_dir,
        },
        'grid': {
            'coordinates': 'coordinates_seagrid_SalishSea.nc',
            'bathymetry': 'bathy_meter_SalishSea2.nc',
        },
        'output': {
            'files': 'iodef.xml',
        }
    }
    if nemo34:
        if forcing is None:
            run_description['forcing'] = {
                'atmospheric':
                '/results/forcing/atmospheric/GEM2.5/operational/',
                'initial conditions': init_conditions,
                'open boundaries': 'open_boundaries/',
                'rivers': 'rivers/',
            }
        else:
            run_description['forcing'] = forcing
        if namelists is None:
            run_description['namelists'] = [
                'namelist.time',
                'namelist.domain',
                'namelist.surface',
                'namelist.lateral',
                'namelist.bottom',
                'namelist.tracers',
                'namelist.dynamics',
                'namelist.compute',
            ]
        else:
            run_description['namelists'] = namelists
    else:
        run_description['paths']['XIOS'] = XIOS_code
        run_description['forcing'] = forcing
        if namelists is None:
            run_description['namelists'] = {
                'namelist_cfg': [
                    'namelist.time',
                    'namelist.domain',
                    'namelist.surface',
                    'namelist.lateral',
                    'namelist.bottom',
                    'namelist.tracer',
                    'namelist.dynamics',
                    'namelist.vertical',
                    'namelist.compute',
                ]
            }
        else:
            run_description['namelists'] = namelists
        run_description['output'] = {
            'domain': 'domain_def.xml',
            'fields': None,
            'separate XIOS server': True,
            'XIOS servers': 1,
        }
        if NEMO_code is not None:
            run_description['output']['fields'] = os.path.join(
                NEMO_code, 'NEMOGCM/CONFIG/SHARED/field_def.xml'
            )
    return run_description


def run_in_subprocess(run_id, run_desc, results_dir, nemo34=False):
    """Execute `nemo run` in a subprocess.

    :arg str run_id: Job identifier that appears in the :command:`qstat`
                     listing.
                     A temporary run description YAML file is created
                     with the name :file:`{run_id}_subprocess_run.yaml`.

    :arg dict run_desc: Run description data structure that will be
                        written to the temporary YAML file.

    :arg boolean nemo34: Execute a NEMO-3.4 run;
                         the default is to execute a NEMO-3.6 run

    :arg results_dir: Directory to store results into.
    :type results_dir: str
    """
    yaml_file = '{}_subprocess_run.yaml'.format(run_id)
    with open(yaml_file, 'wt') as f:
        yaml.dump(run_desc, f, default_flow_style=False)
    cmd = ['salishsea', 'run']
    if nemo34:
        cmd.append('--nemo3.4')
    cmd.extend([yaml_file, results_dir])
    try:
        output = subprocess.check_output(
            cmd, stderr=subprocess.STDOUT, universal_newlines=True
        )
        for line in output.splitlines():
            if line:
                log.info(line)
    except subprocess.CalledProcessError as e:
        log.error(
            'subprocess {cmd} failed with return code {status}'.
            format(cmd=cmd, status=e.returncode)
        )
        for line in e.output.splitlines():
            if line:
                log.error(line)
    os.unlink(yaml_file)


def _run_subcommand(app, app_args, argv):
    """Run a sub-command with argv as arguments via its plug-in
    interface.

    Based on :py:meth:`cliff.app.run_subcommand`.

    :arg app: Application instance invoking the command.
    :type app: :py:class:`cliff.app.App`

    :arg app_args: Application arguments.
    :type app_args: :py:class:`argparse.Namespace`

    :arg argv: Sub-command arguments.
    :type argv: list
    """
    command_manager = cliff.commandmanager.CommandManager(
        'salishsea.app', convert_underscores=False
    )
    try:
        subcommand = command_manager.find_command(argv)
    except ValueError as err:
        if app_args.debug:
            raise
        else:
            log.error(err)
        return 2
    cmd_factory, cmd_name, sub_argv = subcommand
    cmd = cmd_factory(app, app_args)
    try:
        cmd_parser = cmd.get_parser(cmd_name)
        parsed_args = cmd_parser.parse_args(sub_argv)
        result = cmd.take_action(parsed_args)
    except Exception as err:
        result = 1
        if app_args.debug:
            log.exception(err)
        else:
            log.error(err)
    return result


def pbs_common(
    run_description,
    n_processors,
    email,
    results_dir,
    pmem='2000mb',
):
    """Return the common PBS directives used to run NEMO in a TORQUE/PBS
    multiple processor context.

    The string that is returned is intended for inclusion in a bash script
    that will be to the TORQUE/PBS queue manager via the :command:`qsub`
    command.

    :arg dict run_description: Run description data structure.

    :arg int n_processors: Number of processors that the run will be
                           executed on.
                           For NEMO-3.6 runs this is the sum of NMEO and
                           XIOS processors.

    :arg str email: Email address to send job begin, end & abort
                    notifications to.

    :arg results_dir: Directory to store results into.
    :type results_dir: str or :py:class:`pathlib.Path`

    :arg str pmem: Memory per processor.

    :returns: PBS directives for run script.
    :rtype: Unicode str
    """
    try:
        td = datetime.timedelta(seconds=run_description['walltime'])
    except TypeError:
        t = datetime.datetime.strptime(
            run_description['walltime'], '%H:%M:%S'
        ).time()
        td = datetime.timedelta(
            hours=t.hour, minutes=t.minute, seconds=t.second
        )
    walltime = td2hms(td)
    pbs_directives = (
        u'#PBS -N {run_id}\n'
        u'#PBS -S /bin/bash\n'
        u'#PBS -l procs={procs}\n'
        u'# memory per processor\n'
        u'#PBS -l pmem={pmem}\n'
        u'#PBS -l walltime={walltime}\n'
        u'# email when the job [b]egins and [e]nds, or is [a]borted\n'
        u'#PBS -m bea\n'
        u'#PBS -M {email}\n'
        u'# stdout and stderr file paths/names\n'
        u'#PBS -o {results_dir}/stdout\n'
        u'#PBS -e {results_dir}/stderr\n'
    ).format(
        run_id=run_description['run_id'],
        procs=n_processors,
        pmem=pmem,
        walltime=walltime,
        email=email,
        results_dir=results_dir,
    )
    return pbs_directives


def td2hms(timedelta):
    """Return a string that is the timedelta value formated as H:M:S
    with leading zeros on the minutes and seconds values.

    :arg timedelta: Time interval to format.
    :type timedelta: :py:obj:datetime.timedelta

    :returns: H:M:S string with leading zeros on the minutes and seconds
              values.
    :rtype: unicode
    """
    seconds = int(timedelta.total_seconds())
    periods = (('hour', 60 * 60), ('minute', 60), ('second', 1))
    hms = []
    for period_name, period_seconds in periods:
        period_value, seconds = divmod(seconds, period_seconds)
        hms.append(period_value)
    return u'{0[0]}:{0[1]:02d}:{0[2]:02d}'.format(hms)
